### PR TITLE
docs: autogenerate flow variables table

### PR DESCRIPTION
**Pull request by oharboe**

### CODE REVIEW

**Comment by oharboe** [comment]
@maliberty Something like this will be a first pass, but I need to flesh out the metadata a bit to get back to more or less where FlowVariables.md used to be at.

File: flow/scripts/variables.yaml [hunk 1]
```diff
@@ -0,0 +1,577 @@
...
+POST_CTS_TCL:
+  description: >
+    Specifies a Tcl script with commands to run after CTS is completed.
+  stages:
+    - cts
+FASTROUTE_TCL:
+  description: >
+    Specifies a Tcl script with commands to run before FastRoute.
+  stages:
+    - grt
```
Comment (maliberty) [inline flow/scripts/variables.yaml]: I thought you would leave the stages as all by default.  This is needed in global_place.tcl too (and gui.tcl).

Comment (oharboe) [reply flow/scripts/variables.yaml]: yes, but I was looking at the documentation sections and pondering how to best represent things...

Comment (oharboe) [reply flow/scripts/variables.yaml]: KISS: refine stages later and leave variables in "all" section when stages section is not carefully curated.

**Comment by oharboe** [comment]
@maliberty Thoughts? 

ci pr-merge seems unrelated.

**Comment by oharboe** [comment]
@maliberty Flaky autotuner test in pr-merge? It went away last time with a retriggered build?

File: flow/scripts/variables.py
Comment (maliberty) [inline flow/scripts/variables.py:1]: Needs a general comment about its purpose.  The file name is not very descriptive (it generates documentation which you couldn't guess).

File: flow/scripts/stage_variables.py
Comment (maliberty) [inline flow/scripts/stage_variables.py:1]: This is the set of variables *not* in the given stage.  The filename gives the opposite impression.

**Review by maliberty — CHANGES_REQUESTED** [review]

**Comment by maliberty** [comment]
image

is not very readable and also misleading (eg it suggests synthesis depends on the KLAYOUT_TECH_FILE which it clearly doesn't).

**Comment by oharboe** [comment]
> image
> 
> is not very readable and also misleading (eg it suggests synthesis depends on the KLAYOUT_TECH_FILE which it clearly doesn't).

Fixed misleading bit: only list variables if they have the stage in "stages:".

As to more readable...

Suggestions?

This is a summary of variables for that stage.

**Comment by maliberty** [comment]
> Fixed misleading bit: only list variables if they have the stage in "stages:".

Did you push this?

File: flow/scripts/variables.py
Comment (oharboe) [reply flow/scripts/variables.py:1]: fixed

File: flow/scripts/stage_variables.py
Comment (oharboe) [reply flow/scripts/stage_variables.py:1]: fixed

**Comment by oharboe** [comment]
> > Fixed misleading bit: only list variables if they have the stage in "stages:".
> 
> Did you push this?

No, I'm writing up comments as I fix, then I will push. There's no great workflow for where to keep such comments in "draft" until I have pushed...

**Comment by oharboe** [comment]
@maliberty As to when re-review: I retrigger a review after updating the PR, so I guess that is the best that I know of w.r.t. the review being ready for another review. Generally ignore noise from github until then...

File: flow/scripts/gui-preload.tcl
Comment (maliberty) [inline flow/scripts/gui-preload.tcl:1]: Not part of this PR?

Comment (oharboe) [reply flow/scripts/gui-preload.tcl:1]: accident. Delete and force push or revert?

**Comment by maliberty** [comment]
I think generating a bullet list of variables for each stage would be more readable.  If they could link back to their definition that would be helpful.  

I would be nice to have a list of variables that are "all" and those that are uncategorized.  It will be easier to push those forward if there is a list.

**Comment by maliberty** [comment]
Once this is all settled I'll run a full CI as the erased envars need testing in all flows.

**Comment by oharboe** [comment]
> I think generating a bullet list of variables for each stage would be more readable. 

That would give you doom scrolling, no?

File: flow/scripts/gui-preload.tcl
Comment (oharboe) [reply flow/scripts/gui-preload.tcl:1]: deleted by force push.

**Comment by oharboe** [comment]
> I think generating a bullet list of variables for each stage would be more readable. If they could link back to their definition that would be helpful.
> 
> I would be nice to have a list of variables that are "all" and those that are uncategorized. It will be easier to push those forward if there is a list.

Fixed

**Comment by maliberty** [comment]
Secure CI started.

**Comment by maliberty** [comment]
The CI is fine but the lint error needs fixing.

**Comment by oharboe** [comment]
> The CI is fine but the lint error needs fixing.

Fixed.

**Comment by oharboe** [comment]
I prefer to have this merged before I update docs in other PRs....
